File: app/src/main/java/org/stepic/droid/ui/activities/RegisterActivity.java
package org.stepic.droid.ui.activities;

import android.app.ProgressDialog;
import android.os.Bundle;
import android.support.design.widget.TextInputLayout;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.KeyEvent;
import android.view.View;
import android.view.inputmethod.EditorInfo;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.google.firebase.analytics.FirebaseAnalytics;
import com.squareup.okhttp.ResponseBody;

import org.jetbrains.annotations.Nullable;
import org.stepic.droid.R;
import org.stepic.droid.analytic.Analytic;
import org.stepic.droid.base.FragmentActivityBase;
import org.stepic.droid.core.ActivityFinisher;
import org.stepic.droid.core.ProgressHandler;
import org.stepic.droid.util.ProgressHelper;
import org.stepic.droid.util.ValidatorUtil;
import org.stepic.droid.web.RegistrationResponse;

import java.lang.annotation.Annotation;

import butterknife.BindView;
import butterknife.BindString;
import butterknife.ButterKnife;
import butterknife.OnFocusChange;
import retrofit.Callback;
import retrofit.Converter;
import retrofit.Response;
import retrofit.Retrofit;


public class RegisterActivity extends FragmentActivityBase {

    public static final String ERROR_DELIMITER = " ";

    @BindView(R.id.root_view)
    View mRootView;

    @BindView(R.id.sign_up_btn)
    Button mCreateAccountButton;

    @BindView(R.id.actionbar_close_btn_layout)
    View mCloseButton;

    @BindView(R.id.first_name_reg)
    TextView mFirstNameView;

    @BindView(R.id.second_name_reg)
    TextView mSecondNameView;

    @BindView(R.id.email_reg)
    TextView mEmailView;

    @BindView(R.id.password_reg)
    TextView mPassword;

    @BindView(R.id.first_name_reg_wrapper)
    TextInputLayout mFirstNameViewWrapper;

    @BindView(R.id.second_name_reg_wrapper)
    TextInputLayout mSecondNameViewWrapper;

    @BindView(R.id.email_reg_wrapper)
    TextInputLayout mEmailViewWrapper;

    @BindView(R.id.password_reg_wrapper)
    TextInputLayout mPasswordWrapper;

    @BindString(R.string.password_too_short)
    String mPasswordTooShortMessage;


    ProgressDialog mProgress;
    TextWatcher mPasswordWatcher;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(org.stepic.droid.R.layout.activity_register);
        unbinder = ButterKnife.bind(this);
        overridePendingTransition(org.stepic.droid.R.anim.slide_in_from_bottom, org.stepic.droid.R.anim.no_transition);

        hideSoftKeypad();

        mCreateAccountButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                createAccount();
            }
        });
        mCloseButton.setOnClickListener(new View.OnClickListener() {

            @Override
            public void onClick(View v) {
                finish();
            }
        });

        mProgress = new ProgressDialog(this);
        mProgress.setTitle(getString(R.string.loading));
        mProgress.setMessage(getString(R.string.loading_message));
        mProgress.setCancelable(false);

        mPassword.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                boolean handled = false;
                if (actionId == EditorInfo.IME_ACTION_SEND) {
                    createAccount();
                    handled = true;
                }
                return handled;
            }
        });

        mPassword.addTextChangedListener(mPasswordWatcher = new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {

            }

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {

            }

            @Override
            public void afterTextChanged(Editable s) {
                if (ValidatorUtil.isPasswordLengthValid(s.length())) {
                    hideError(mPasswordWrapper);
                }
            }
        });

        mRootView.requestFocus();
    }


    private void createAccount() {
        String firstName = mFirstNameView.getText().toString().trim();
        String lastName = mSecondNameView.getText().toString().trim();
        final String email = mEmailView.getText().toString().trim();
        final String password = mPassword.getText().toString();

        analytic.reportEvent(Analytic.Interaction.CLICK_REGISTER_BUTTON);

        boolean isOk = true;

        if (!ValidatorUtil.isPasswordValid(password)) {
            showError(mPasswordWrapper, mPasswordTooShortMessage);
            isOk = false;
        }

        if (isOk) {
            hideError(mFirstNameViewWrapper);
            hideError(mSecondNameViewWrapper);
            hideError(mEmailViewWrapper);
            hideError(mPasswordWrapper);

            mShell.getApi().signUp(firstName, lastName, email, password).enqueue(new Callback<RegistrationResponse>() {
                @Override
                public void onResponse(Response<RegistrationResponse> response, Retrofit retrofit) {
                    ProgressHelper.dismiss(mProgress);
                    if (response.isSuccess()) {
                        analytic.reportEvent(FirebaseAnalytics.Event.SIGN_UP);
                        mLoginManager.login(email, password, new ProgressHandler() {
                            @Override
                            public void activate() {
                                hideSoftKeypad();
                                ProgressHelper.activate(mProgress);
                            }

                            @Override
                            public void dismiss() {
                                ProgressHelper.dismiss(mProgress);
                            }
                        }, new ActivityFinisher() {
                            @Override
                            public void onFinish() {
                                finish();
                            }
                        });
                    } else {
                        Converter<ResponseBody, RegistrationResponse> errorConverter =
                                retrofit.responseConverter(RegistrationResponse.class, new Annotation[0]);
                        RegistrationResponse error = null;
                        try {
                            error = errorConverter.convert(response.errorBody());
                        } catch (Exception e) {
                            analytic.reportError(Analytic.Error.REGISTRATION_IMPORTANT_ERROR, e); //it is unknown response Expected BEGIN_OBJECT but was STRING at line 1 column 1 path
                        }
                        handleErrorRegistrationResponse(error);
                    }

                }

                @Override
                public void onFailure(Throwable t) {
                    ProgressHelper.dismiss(mProgress);
                    Toast.makeText(RegisterActivity.this, R.string.connectionProblems, Toast.LENGTH_SHORT).show();
                }
            });
        }
    }

    @Override
    protected void onStart() {
        super.onStart();
        bus.register(this);
    }

    @Override
    protected void onStop() {
        super.onStop();
        bus.unregister(this);
    }

    @Override
    protected void onDestroy() {
        mPassword.removeTextChangedListener(mPasswordWatcher);
        mPassword.setOnEditorActionListener(null);
        super.onDestroy();
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(org.stepic.droid.R.anim.no_transition, org.stepic.droid.R.anim.slide_out_to_bottom);
    }

    private void hideError(TextInputLayout textInputLayout) {
        if (textInputLayout != null) {
            textInputLayout.setError("");
            textInputLayout.setErrorEnabled(false);
        }
    }

    private void showError(TextInputLayout textInputLayout, String errorText) {
        if (textInputLayout != null) {
            textInputLayout.setErrorEnabled(true);
            textInputLayout.setError(errorText);
        }
    }

    private void handleErrorRegistrationResponse(@Nullable RegistrationResponse registrationResponse) {
        if (registrationResponse == null) return;
        showError(mEmailViewWrapper, getErrorString(registrationResponse.getEmail()));
        showError(mFirstNameViewWrapper, getErrorString(registrationResponse.getFirst_name()));
        showError(mSecondNameViewWrapper, getErrorString(registrationResponse.getLast_name()));
        showError(mPasswordWrapper, getErrorString(registrationResponse.getPassword()));
    }

    @Nullable
    private String getErrorString(String[] values) {
        if (values == null || values.length == 0) return null;
        StringBuilder sb = new StringBuilder();
        for (int i = 0; i < values.length; i++) {
            sb.append(values[i]);
            if (i != values.length - 1) {
                sb.append(ERROR_DELIMITER);
            }
        }
        return sb.toString();
    }

    @OnFocusChange({R.id.email_reg, R.id.first_name_reg, R.id.second_name_reg})
    public void setClearErrorOnFocus(View view, boolean hasFocus) {
        if (hasFocus) {
            if (view.getId() == R.id.email_reg) {
                hideError(mEmailViewWrapper);
            }

            if (view.getId() == R.id.first_name_reg) {
                hideError(mFirstNameViewWrapper);
            }
            if (view.getId() == R.id.second_name_reg) {
                hideError(mSecondNameViewWrapper);
            }
        }
    }

}

File: app/src/main/java/org/stepic/droid/ui/activities/RegisterActivity.kt
package org.stepic.droid.ui.activities

import android.os.Bundle
import android.text.Editable
import android.text.Spannable
import android.text.SpannableString
import android.text.TextWatcher
import android.text.method.LinkMovementMethod
import android.view.View
import android.view.inputmethod.EditorInfo
import android.widget.Toast
import androidx.appcompat.app.AppCompatDelegate
import androidx.core.content.res.ResourcesCompat
import com.google.android.gms.auth.api.credentials.Credential
import kotlinx.android.synthetic.main.activity_register.*
import okhttp3.ResponseBody
import org.stepic.droid.R
import org.stepic.droid.analytic.Analytic
import org.stepic.droid.analytic.LoginInteractionType
import org.stepic.droid.base.App
import org.stepic.droid.core.LoginFailType
import org.stepic.droid.core.presenters.LoginPresenter
import org.stepic.droid.core.presenters.contracts.LoginView
import org.stepic.droid.model.Credentials
import org.stepic.droid.ui.dialogs.LoadingProgressDialog
import org.stepic.droid.ui.util.setOnKeyboardOpenListener
import org.stepic.droid.util.ProgressHelper
import org.stepic.droid.util.ValidatorUtil
import org.stepic.droid.util.getMessageFor
import org.stepic.droid.util.stripUnderlinesFromLinks
import org.stepic.droid.util.toBundle
import org.stepic.droid.web.Api
import org.stepic.droid.web.RegistrationResponse
import org.stepik.android.view.base.ui.span.TypefaceSpanCompat
import retrofit2.Retrofit
import ru.nobird.android.view.base.ui.extension.hideKeyboard
import javax.inject.Inject

class RegisterActivity : SmartLockActivityBase(), LoginView {
    companion object {
        const val ERROR_DELIMITER = " "
        const val TAG = "RegisterActivity"

        init {
            AppCompatDelegate.setCompatVectorFromResourcesEnabled(true)
        }
    }

    private val passwordTooShortMessage by lazy {
        resources.getString(R.string.password_too_short)
    }
    private val termsMessageHtml by lazy {
        resources.getString(R.string.terms_message_register)
    }

    private lateinit var progressBar: LoadingProgressDialog

    @Inject
    lateinit var loginPresenter: LoginPresenter

    @Inject
    internal lateinit var retrofit: Retrofit

    @Inject
    internal lateinit var api: Api

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_register)
        App.componentManager().loginComponent(TAG).inject(this)

        initTitle()

        termsPrivacyRegisterTextView.movementMethod = LinkMovementMethod.getInstance()
        termsPrivacyRegisterTextView.text = textResolver.fromHtml(termsMessageHtml)
        stripUnderlinesFromLinks(termsPrivacyRegisterTextView)

        signUpButton.setOnClickListener { createAccount(LoginInteractionType.button) }

        progressBar = LoadingProgressDialog(this)

        passwordField.setOnEditorActionListener { _, actionId, _ ->
            var handled = false
            if (actionId == EditorInfo.IME_ACTION_SEND) {
                analytic.reportEvent(Analytic.Registration.CLICK_SEND_IME)
                createAccount(LoginInteractionType.ime)
                handled = true
            }
            handled
        }
        val reportAnalyticWhenTextBecomeNotBlank = object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {
                if (s.isNullOrBlank()) {
                    analytic.reportEvent(Analytic.Registration.TYPING_TEXT_FIELDS)
                }
            }

            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {
            }

            override fun afterTextChanged(s: Editable?) {
                onClearError()
                setSignUpButtonState()
            }
        }

        firstNameField.addTextChangedListener(reportAnalyticWhenTextBecomeNotBlank)
        emailField.addTextChangedListener(reportAnalyticWhenTextBecomeNotBlank)
        passwordField.addTextChangedListener(reportAnalyticWhenTextBecomeNotBlank)

        val onFocusChangeListener = View.OnFocusChangeListener { _, hasFocus ->
            reportTap(hasFocus)
        }
        firstNameField.onFocusChangeListener = onFocusChangeListener
        emailField.onFocusChangeListener = onFocusChangeListener
        passwordField.onFocusChangeListener = onFocusChangeListener

        firstNameField.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_NEXT) {
                emailField.requestFocus()
                true
            } else {
                false
            }
        }

        emailField.setOnEditorActionListener { _, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_NEXT) {
                passwordField.requestFocus()
                true
            } else {
                false
            }
        }

        registerRootView.requestFocus()

        initGoogleApiClient()

        setSignUpButtonState()

        setOnKeyboardOpenListener(root_view, {
            stepikLogo.visibility = View.GONE
            signUpText.visibility = View.GONE
        }, {
            stepikLogo.visibility = View.VISIBLE
            signUpText.visibility = View.VISIBLE
        })

        loginPresenter.attachView(this)
    }

    private fun setSignUpButtonState() {
        signUpButton.isEnabled = emailField.text.isNotBlank() && firstNameField.text.isNotBlank() && passwordField.text.isNotBlank()
    }

    private fun initTitle() {
        val signUpString = getString(R.string.sign_up)
        val signUpSuffix = getString(R.string.sign_up_with_email_suffix)

        val spannableSignIn = SpannableString(signUpString + signUpSuffix)
        val typeface = ResourcesCompat.getFont(this, R.font.roboto_medium)

        spannableSignIn.setSpan(TypefaceSpanCompat(typeface), 0, signUpString.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)

        signUpText.text = spannableSignIn
    }

    private fun reportTap(hasFocus: Boolean) {
        if (hasFocus) {
            analytic.reportEvent(Analytic.Registration.TAP_ON_FIELDS)
        }
    }


    private fun createAccount(interactionType: LoginInteractionType) {
        analytic.reportEvent(Analytic.Registration.CLICK_WITH_INTERACTION_TYPE, interactionType.toBundle())
        currentFocus?.hideKeyboard()

        val firstName = firstNameField.text.toString().trim()
        val lastName = " " // registrationSecondName.text.toString().trim()
        val email = emailField.text.toString().trim()
        val password = passwordField.text.toString()

        analytic.reportEvent(Analytic.Interaction.CLICK_REGISTER_BUTTON)

        var isOk = true

        if (!ValidatorUtil.isPasswordValid(password)) {
            showError(passwordTooShortMessage)
            isOk = false
        }

        if (isOk) {
            onLoadingWhileLogin()
            loginPresenter.signUp(firstName, lastName, email, password)
//            api.signUp(firstName, lastName, email, password).enqueue(object : Callback<RegistrationResponse> {
//                override fun onResponse(call: Call<RegistrationResponse>, response: Response<RegistrationResponse>) {
//                    if (response.isSuccessful) {
//                        analytic.reportEvent(FirebaseAnalytics.Event.SIGN_UP)
//
//                        loginPresenter.login(email, password, isAfterRegistration = true)
//                    } else {
//                        ProgressHelper.dismiss(progressBar)
//                        response.errorBody()
//                        val errorConverter = retrofit.responseBodyConverter<RegistrationResponse>(RegistrationResponse::class.java, arrayOfNulls<Annotation>(0))
//                        var error: RegistrationResponse? = null
//                        try {
//                            error = errorConverter.convert(response.errorBody()!!)
//                        } catch (e: Exception) {
//                            analytic.reportError(Analytic.Error.REGISTRATION_IMPORTANT_ERROR, e) //it is unknown response Expected BEGIN_OBJECT but was STRING at line 1 column 1 path
//                        }
//
//                        handleErrorRegistrationResponse(error)
//                    }
//
//                }
//
//                override fun onFailure(call: Call<RegistrationResponse>, t: Throwable) {
//                    ProgressHelper.dismiss(progressBar)
//                    Toast.makeText(this@RegisterActivity, R.string.connectionProblems, Toast.LENGTH_SHORT).show()
//                }
//            })
        }
    }

    override fun onRegistrationFailed(responseBody: ResponseBody?) {
        ProgressHelper.dismiss(progressBar)
        val errorConverter = retrofit.responseBodyConverter<RegistrationResponse>(RegistrationResponse::class.java, arrayOfNulls<Annotation>(0))
        var error: RegistrationResponse? = null
        try {
            error = errorConverter.convert(responseBody!!)
        } catch (e: Exception) {
            analytic.reportError(Analytic.Error.REGISTRATION_IMPORTANT_ERROR, e) //it is unknown response Expected BEGIN_OBJECT but was STRING at line 1 column 1 path
        }
        handleErrorRegistrationResponse(error)
    }

    private fun onClearError() {
        signUpButton.isEnabled = true
        registerForm.isEnabled = true
        registerErrorMessage.visibility = View.GONE
    }

    override fun onDestroy() {
        loginPresenter.detachView(this)
        passwordField.setOnEditorActionListener(null)
        if (isFinishing) {
            App.componentManager().releaseLoginComponent(TAG)
        }
        super.onDestroy()
    }

    override fun applyTransitionPrev() {} // we need default system animation

    private fun showError(errorText: String?) {
        errorText?.let {
            analytic.reportEventWithName(Analytic.Registration.ERROR, errorText)
            if (registerErrorMessage.visibility == View.GONE) {
                signUpButton.isEnabled = false
                registerForm.isEnabled = false
                registerErrorMessage.text = it
                registerErrorMessage.visibility = View.VISIBLE
            }
        }
    }

    private fun handleErrorRegistrationResponse(registrationResponse: RegistrationResponse?) {
        if (registrationResponse == null) return
        showError(getErrorString(registrationResponse.email))
        showError(getErrorString(registrationResponse.first_name))
        showError(getErrorString(registrationResponse.last_name))
        showError(getErrorString(registrationResponse.password))
    }

    private fun getErrorString(values: Array<String?>?): String? =
        values
            ?.takeIf { it.isNotEmpty() }
            ?.joinToString(separator = ERROR_DELIMITER)

    override fun onFailLogin(type: LoginFailType, credential: Credential?) {
        ProgressHelper.dismiss(progressBar)
        Toast.makeText(this, getMessageFor(type), Toast.LENGTH_SHORT).show()
    }

    override fun onSuccessLogin(credentials: Credentials?) {
        ProgressHelper.dismiss(progressBar)
        if (credentials == null || !checkPlayServices() || googleApiClient?.isConnected != true) {
            openMainFeed()
        } else {
            //only if we have not null data (we can apply smart lock && google api client is connected and available
            requestToSaveCredentials(credentials)
        }
    }

    override fun onCredentialSaved() = openMainFeed()

    private fun openMainFeed() {
        screenManager.showMainFeedAfterLogin(this, courseFromExtra)
    }

    override fun onLoadingWhileLogin() {
        currentFocus?.hideKeyboard()
        ProgressHelper.activate(progressBar)
    }

    override fun onSocialLoginWithExistingEmail(email: String) {}

}
